### PR TITLE
feat: trigger completed actions

### PR DESCRIPTION
**Pull request by sshanzel**
Retro checks for completed action:
- Have taken the same path as we did on the markdown content update.
- One, it is much easier to develop, as it is more predictable running functions.
- Also testable as shown in the newly introduced test suite.

Completed actions implemented:
- create_squad
- join_squad
- squad_first_comment
- squad_first_post
- squad_invite
- my_feed

Actions not included in this PR:
- edit_welcome_post: will happen on the edit welcome post mutation.
- browser_extension: will happen onload at FE to send mutation to complete the action
- enabled_notification: there is currently no API on onesignal/node to fetch by user id. Fetching the devices is also limited to 300 rows per request. We will have to send a mutation once detected in FE.

### CODE REVIEW

File: src/schema/feeds.ts
Comment (sshanzel) [inline src/schema/feeds.ts:1]: @rebelchris let me know what you think of the changes here. Or do you want to have a subscription on the alerts myFeed property changes?

File: src/entity/UserAction.ts
Comment (rebelchris) [inline src/entity/UserAction.ts:4]: This one is wrong in this version right? We described it as `enabled_notification`

File: src/schema/actions.ts
Comment (rebelchris) [inline src/schema/actions.ts:52]: Gotta love it!

File: src/schema/feeds.ts
Comment (rebelchris) [reply src/schema/feeds.ts:1]: This seems like it could trigger many, many times?
Not sure on the implication maybe the less friction for now would be listening on alerts.

Not particular against this solution, but maybe let's ask @idoshamun if he sees any issue with this?

File: src/workers/notifications/postAdded.ts
Comment (rebelchris) [inline src/workers/notifications/postAdded.ts:80]: Don't think this one is complete.
We would need to check inside this squad scope as well right?

Comment (rebelchris) [reply src/workers/notifications/postAdded.ts:80]: Maybe we should also explicitly check for squad type post as they might be author of old articles?

File: src/workers/notifications/postAdded.ts
Comment (rebelchris) [inline src/workers/notifications/postAdded.ts:82]: This would just be a "Extra"? 
Because technically they could delete a post, then post again (I do know ignore will trigger)

File: src/workers/notifications/utils.ts
Comment (rebelchris) [inline src/workers/notifications/utils.ts:74]: Would it not be simpler to always try the insert or ignore directly?
Seems we don't particular would care about querying first right?

Also the query doesn't take specific squads in concern

**Review by rebelchris — CHANGES_REQUESTED** [review]
Wow love the approach, have some minor suggestion but like the way this is going.

File: src/schema/feeds.ts
Comment (sshanzel) [reply src/schema/feeds.ts:1]: Good point on the frequency which would make it somehow ideal to be listening on alerts.

File: src/entity/UserAction.ts
Comment (sshanzel) [reply src/entity/UserAction.ts:4]: Yes, this should be fixed in the PR where we utilized this. Was just a placeholder initially.

File: src/workers/notifications/postAdded.ts
Comment (sshanzel) [reply src/workers/notifications/postAdded.ts:80]: Though we check the Source Type to be Squad, I agree having to check the post type would be best.

File: src/workers/notifications/postAdded.ts
Comment (sshanzel) [reply src/workers/notifications/postAdded.ts:82]: Oh, but when they delete a post we just set the property `deleted = true` right?

File: src/workers/notifications/utils.ts
Comment (sshanzel) [reply src/workers/notifications/utils.ts:74]: Yeah, was also thinking that initially but hesitated. Now that you mentioned it, let's do that.

File: src/workers/notifications/postAdded.ts
Comment (rebelchris) [reply src/workers/notifications/postAdded.ts:80]: Yeah we do, but this would just get any posts count right? 👀
So the count could be different than what you want.

Comment (sshanzel) [reply src/workers/notifications/postAdded.ts:80]: Totally right!

File: src/workers/notifications/postAdded.ts
Comment (sshanzel) [reply src/workers/notifications/postAdded.ts:83]: Indeed. Should be available on the condition now 🥳

**Comment by sshanzel** [comment]
@rebelchris have now added the retro check and its tests.

**Comment by rebelchris** [comment]
@sshanzel For the enable_notifications one idea I have is that we can actually export a csv of all current used_id in OneSignal and add manual query for those once we go live? (If it allows us to have less FE overhead?) 
Would love to hear what you and @idoshamun think about this.

**Comment by sshanzel** [comment]
> @sshanzel For the enable_notifications one idea I have is that we can actually export a csv of all current used_id in OneSignal and add manual query for those once we go live? (If it allows us to have less FE overhead?)
Would love to hear what you and @idoshamun think about this.

Great idea. I think we can work with their exported CSV.

**Comment by idoshamun** [comment]
> @sshanzel For the enable_notifications one idea I have is that we can actually export a csv of all current used_id in OneSignal and add manual query for those once we go live? (If it allows us to have less FE overhead?) Would love to hear what you and @idoshamun think about this.

Great! It's easy to achieve actually

File: bin/completedUserActions.ts
Comment (idoshamun) [inline bin/completedUserActions.ts:94]: Instead of selecting and then inserting, why don't you use insert from select?
You will not have to deal with streaming or anything like that and it's much more reliable

File: src/entity/UserAction.ts
Comment (idoshamun) [inline src/entity/UserAction.ts:4]: Let's keep the same convention

```suggestion
  Notification = 'enable_notification',
```

File: bin/completedUserActions.ts
Comment (sshanzel) [reply bin/completedUserActions.ts:94]: Oh, cool. Just thought of doing the same process as the past executions, let's do that 🥳

File: src/schema/feeds.ts
Comment (idoshamun) [reply src/schema/feeds.ts:1]: You don't want to add anything unnecessary for the feed call. It may have a tremendous impact.
We already have a mechanism to identify when the user enable my feed, let's use it.

File: src/workers/notifications/postAdded.ts
Comment (idoshamun) [reply src/workers/notifications/postAdded.ts:83]: I think we discussed in the other PR that we don't need to count the posts if we add retroactively the actions

File: src/schema/feeds.ts
Comment (sshanzel) [reply src/schema/feeds.ts:1]: Yep. I just realized we used the `alerts.filter` for this exact case, when the user updated their filters. We can probably do the inline direct from the updateAlerts mutation.

File: src/workers/notifications/squadMemberJoined.ts
Comment (idoshamun) [inline src/workers/notifications/squadMemberJoined.ts:35]: This is a bit more tricky than that, because now admins can promote other admins.
And these new admins didn't really create a squad

**Review by idoshamun — CHANGES_REQUESTED** [review]
Great start but still some work to do

File: src/workers/notifications/postAdded.ts
Comment (sshanzel) [reply src/workers/notifications/postAdded.ts:83]: Yes, I will rebase this once the PR is merged.

File: src/workers/notifications/squadMemberJoined.ts
Comment (sshanzel) [reply src/workers/notifications/squadMemberJoined.ts:35]: Yes, though it was discussed to provide the checklist for owners even if it came from a promotion.

Sab mentioned:

> We need to understand that this means if a user becomes a Squad admin (owner) then they see the owner checklist and it replaces any older checklists that may or may not have been completed

https://dailydotdev.slack.com/archives/C0506KU4RH9/p1681985308241999?thread_ts=1681978406.223319&cid=C0506KU4RH9

Another relevant discussion:

https://dailydotdev.slack.com/archives/C0506KU4RH9/p1683019582216549

Comment (idoshamun) [reply src/workers/notifications/squadMemberJoined.ts:35]: Ok!

File: src/schema/alerts.ts
Comment (sshanzel) [inline src/schema/alerts.ts:114]: @rebelchris though we agreed on using `my_feed` but I realized `alert.filter` is more accurate and much easier to use as it is a boolean than a string. Do you think we can move forward with this?

**Comment by sshanzel** [comment]
@idoshamun have updated it now to use `insert select` query 🚢 

@rebelchris have created a ticket for doing the OneSignal subscribed users export -> import user_action table.

https://dailydotdev.atlassian.net/jira/software/projects/WT/boards/11/backlog?selectedIssue=WT-1315

File: src/schema/alerts.ts
Comment (rebelchris) [reply src/schema/alerts.ts:114]: What's the difference between the two?
I meant the one we set once someone created my feed (thought it was that one, but open to the other if it's easier and the same)

Comment (rebelchris) [reply src/schema/alerts.ts:114]: Perfect then, this was probably the one I meant initially :)

**Review by rebelchris — APPROVED** [review]
Really neat work here Lee,
Nothing to add from my side, just make double sure the mutations work as they won't be able to revert.

File: bin/completedUserActions.ts
Comment (idoshamun) [inline bin/completedUserActions.ts:18]: I think we should ignore conflicts right?

File: bin/notificationSubscribedUsers.ts
Comment (idoshamun) [inline bin/notificationSubscribedUsers.ts:4]: Please remove the array from this file. It should be provided as a runtime argument.
I don't want to expose user ids in our codebase

Comment (sshanzel) [reply bin/notificationSubscribedUsers.ts:4]: Oh, right. How would we receive the values and what would be the type of the argument? Will it be a comma-separated string?

Comment (idoshamun) [reply bin/notificationSubscribedUsers.ts:4]: I think a path to a file and then we can read the file in runtime

File: bin/completedUserActions.ts
Comment (sshanzel) [reply bin/completedUserActions.ts]: Indeed we should 🚀

File: bin/notificationSubscribedUsers.ts
Comment (sshanzel) [inline bin/notificationSubscribedUsers.ts:4]: WIP 🔨

File: bin/notificationSubscribedUsers.ts
Comment (sshanzel) [inline bin/notificationSubscribedUsers.ts:31]: Used object, for easier filtering unique ids as users can have multiple devices from this CSV.

**Review by idoshamun — APPROVED** [review]
Looks great just one comment and you can merge without my followup approval after validating locally

File: bin/notificationSubscribedUsers.ts
Comment (sshanzel) [reply bin/notificationSubscribedUsers.ts:8]: That makes sense. Just thought of doing it explicitly if possible, but yes, let's just keep it to the ignoring function.

Comment (idoshamun) [reply bin/notificationSubscribedUsers.ts:8]: Just to give you clear guidance, less queries the better. The more you do on the database side the less you have to think about scale

**Comment by sshanzel** [comment]
Oh, and the other one left would be the `squad_invite`. Though we can move forward if we will add another thing to catch user creation. 

Reference for the actions: https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/498401281/Squads+V5#Types-of-action

**Comment by idoshamun** [comment]
@sshanzel what about enable notifications? This will also require frontend side

**Comment by sshanzel** [comment]
> @sshanzel what about enable notifications? This will also require frontend side

Oh, that is right. Though we have a separate runner for the notifications, I agree it would be better to do it in one go. I am just so excited as this is my first time doing this use-case 😂

**Comment by idoshamun** [comment]
Lol yeah. So let's run it once the frontend is done and merged :)

File: bin/notificationSubscribedUsers.ts
Comment (sshanzel) [reply bin/notificationSubscribedUsers.ts]: Thank you for the insight! 🙏

**Pull request merged**
